### PR TITLE
feat: add `utils/any-in-by`

### LINKED ISSUE #823: [RFC]: Add `@stdlib/utils/any-in-by`
### Description

This RFC proposes adding a utility to test whether any "own" and inherited property of a provided object satisfies a predicate function.

This is the `for-in` equivalent of [`@stdlib/utils/any-by`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/any-by).

Package: `@stdlib/utils/any-in-by`
Alias: `anyInBy`

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by itsspriyansh**
Resolves #823 

## Description

> What is the purpose of this pull request?

This pull request:

-   a adds utils/any-in-by package
-   b tests whether any "own" and inherited property of a provided object satisfies a predicate function.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #
-   fixes #

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [ ] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/utils/any-in-by/README.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+<section class="usage">
+
+## Usage
+
+```javascript
+var anyInBy = require( '@stdlib/utils/any-in-by' );
+```
+
+#### anyInBy( collection, predicate\[, thisArg ] )
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/README.md]:
```suggestion
#### anyInBy( object, predicate\[, thisArg ] )
```

File: lib/node_modules/@stdlib/utils/any-in-by/README.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+-   <span class="package-name">[`@stdlib/utils/every-by`][@stdlib/utils/every-by]</span><span class="delimiter">: </span><span class="description">test whether all elements in a collection pass a test implemented by a predicate function.</span>
+-   <span class="package-name">[`@stdlib/utils/for-each`][@stdlib/utils/for-each]</span><span class="delimiter">: </span><span class="description">invoke a function for each element in a collection.</span>
+-   <span class="package-name">[`@stdlib/utils/none-by`][@stdlib/utils/none-by]</span><span class="delimiter">: </span><span class="description">test whether all elements in a collection fail a test implemented by a predicate function.</span>
+-   <span class="package-name">[`@stdlib/utils/some-by`][@stdlib/utils/some-by]</span><span class="delimiter">: </span><span class="description">test whether a collection contains at least `n` elements which pass a test implemented by a predicate function.</span>
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/README.md]: We have tooling to auto-populate this section, so no reason to fill in manually.

```suggestion
```

File: lib/node_modules/@stdlib/utils/any-in-by/README.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+[@stdlib/utils/any-by-right]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/any-by-right
+
+[@stdlib/utils/every-by]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/every-by
+
+[@stdlib/utils/for-each]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/for-each
+
+[@stdlib/utils/none-by]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/none-by
+
+[@stdlib/utils/some-by]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/some-by
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/README.md]: We have tooling to auto-populate these links, so no reason to fill in manually.

```suggestion
```

File: lib/node_modules/@stdlib/utils/any-in-by/README.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+var bool = anyInBy( obj, isPositive );
+// returns true
+```
+
+The invoked `function` is provided three arguments:
+
+-   `value`: collection element
+-   `index`: collection index
+-   `collection`: input collection
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/README.md]:
```suggestion
-   `value`: property value
-   `key`: property key
-   `object`: input object
```

File: lib/node_modules/@stdlib/utils/any-in-by/README.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+<!-- Package usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+
+<section class="notes">
+
+## Notes
+
+-   The `obj` object may contain own properties as well as inherited properties from its prototype chain.
+
+-   If provided an empty collection, the function returns `false`.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/README.md]:
```suggestion
-   If provided an empty object, the function returns `false`.
```

File: lib/node_modules/@stdlib/utils/any-in-by/README.md [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        /* =>
+            0: 1
+            1: undefined
+            2: undefined
+            3: 4
+            4: -1
+        */
+        ```
+
+    -   The function provides limited support for dynamic collections (i.e., collections whose `length` changes during execution).
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/README.md]: Let's remove these notes as specifically about arrays.

```suggestion
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/repl.txt [hunk 7]
```diff
@@ -0,0 +1,41 @@
+{{alias}}( collection, predicate[, thisArg ] )
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/repl.txt]: REPL documentation here needs to be updates as it currently is written for `collection` / `array` inputs, but the function is for `object`s.

File: lib/node_modules/@stdlib/utils/any-in-by/test/test.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+		t.throws( badValue( values ), TypeError, 'throws a type error when provided '+values[i] );
+	}
+	t.end();
+
+	function badValue( value ) {
+		return function badValue() {
+			anyInBy( value, noop );
+		};
+	}
+});
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/test/test.js]: Can remove as duplicate of test in L36-63.

```suggestion
```

However, we need a test for ensuring that function throws an error if provided a predicate argument that is not a function.

File: lib/node_modules/@stdlib/utils/any-in-by/package.json [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      "utility",
+      "utils",
+      "util",
+      "test",
+      "predicate",
+      "any",
+      "array.some",
+      "iterate",
+      "collection",
+      "array-like",
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/package.json]:
```suggestion
      "object",
      "obj",
      "iterate",
```

File: lib/node_modules/@stdlib/utils/any-in-by/lib/main.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+	if ( !isObject( obj ) ) {
+		throw new TypeError( format( 'invalid argument. First argument must be an object. Value: `%s`.', obj ) );
+	}
+	if ( !isFunction( predicate ) ) {
+		throw new TypeError( format( 'invalid argument. Second argument must be a function. Value: `%s`.', predicate ) );
+	}
+
+	for ( key in obj ) {
+		if ( hasOwnProp( obj, key ) && predicate.call( thisArg, obj[key], key, obj ) ) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/lib/main.js]: We should remove the `hasOwnProp( obj, key )` property check as the function is for "own _and inherited_" properties.

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+* }
+*
+* var obj = { 'a': 1, 'b': -2, 'c': 3 };
+*
+* var bool = anyInBy( obj, isNegative );
+* // returns true
+*/
+declare function anyInBy<T = unknown, U = unknown>( obj: { [key: string]: T }, predicate: Predicate<T, U>, thisArg?: ThisParameterType<Predicate<T, U>> ): boolean;
+
+// EXPORTS //
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts:100]: Two empty lines before section header comments:

```suggestion


// EXPORTS //
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+*/
+
+// TypeScript Version: 4.1
+
+/// <reference types="@stdlib/types"/>
+
+import { Collection } from '@stdlib/types/array';
+
+/**
+* Checks whether an element in a collection passes a test.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* Checks whether an object property passes a test.
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+*
+* @returns boolean indicating whether an element in a collection passes a test
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* @returns boolean indicating whether an object property passes a test
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+*/
+type Nullary<U> = ( this: U ) => boolean;
+
+/**
+* Checks whether an element in a collection passes a test.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* Checks whether an object property passes a test.
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+*
+* @param value - collection value
+* @returns boolean indicating whether an element in a collection passes a test
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* @returns boolean indicating whether an object property passes a test
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+*/
+type Unary<T, U> = ( this: U, value: T ) => boolean;
+
+/**
+* Checks whether an element in a collection passes a test.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* Checks whether an object property passes a test.
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+*
+* @returns boolean indicating whether an element in a collection passes a test
+*/
+type Nullary<U> = ( this: U ) => boolean;
+
+/**
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* @param value - property value
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+* @returns boolean indicating whether an element in a collection passes a test
+*/
+type Unary<T, U> = ( this: U, value: T ) => boolean;
+
+/**
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* @param value - property value
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+* @param index - collection index
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* @param key - property key
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+* @returns boolean indicating whether an element in a collection passes a test
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* @returns boolean indicating whether an object property passes a test
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
+* @param index - collection index
+* @param collection - input collection
+* @returns boolean indicating whether an element in a collection passes a test
+*/
+type Ternary<T, U> = ( this: U, value: T, index: number, collection: Collection<T> ) => boolean;
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* Checks whether an object property passes a test.
*
* @param value - property value
* @param key - property key
* @param obj - input object
* @returns boolean indicating whether an object property passes a test
*/
type Ternary<T, U> = ( this: U, value: T, key: string, obj:{ [key: string]: T } ) => boolean;
```

File: lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+* Checks whether an element in a collection passes a test.
+*
+* @param value - collection value
+* @param index - collection index
+* @param collection - input collection
+* @returns boolean indicating whether an element in a collection passes a test
+*/
+type Predicate<T, U> = Nullary<U> | Unary<T, U> | Binary<T, U> | Ternary<T, U>;
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/docs/types/index.d.ts]:
```suggestion
* Checks whether an object property passes a test.
*
* @param value - property value
* @param key - property key
* @param obj - input object
* @returns boolean indicating whether an object property passes a test
*/
type Predicate<T, U> = Nullary<U> | Unary<T, U> | Binary<T, U> | Ternary<T, U>;
```

**Review by Planeshifter — CHANGES_REQUESTED** [review]
Thanks for this PR! Left some comments that will need to be addressed before this can be merged.

**Comment by itsspriyansh** [comment]
@Planeshifter I have resolved all the problems.

File: lib/node_modules/@stdlib/utils/any-in-by/test/test.js [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+// TESTS //
+
+tape( 'main export is a function', function test( t ) {
+	t.ok( true, __filename );
+	t.strictEqual( typeof anyInBy, 'function', 'main export is a function' );
+	t.end();
+});
+
+tape( 'the function throws an error if not provided a object', function test( t ) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/test/test.js]:
```suggestion
tape( 'the function throws an error if not provided an object', function test( t ) {
```

File: lib/node_modules/@stdlib/utils/any-in-by/test/test.js [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+	}
+	t.end();
+
+	function badValue( value ) {
+		return function badValue() {
+			anyInBy( value, noop );
+		};
+	}
+});
+
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/test/test.js:64]: Needs a test for when `predicate` is not a function.

File: lib/node_modules/@stdlib/utils/any-in-by/test/test.js [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+	function underAge( value ) {
+		return ( value < 18 );
+	}
+
+	bool = anyInBy( obj, underAge );
+
+	t.strictEqual( bool, true, 'returns true' );
+	t.end();
+});
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/utils/any-in-by/test/test.js:161]: We should have a test for when the function returns `false`. Our goal is to achieve 100% code coverage, so we should make sure that tests cover all code paths.

**Review by Planeshifter — COMMENTED** [review]
It's coming along well, but I noticed we are still missing some tests. Would you be able to please add these?

**Comment by itsspriyansh** [comment]
@Planeshifter I have fixed the typo and I have added 3 tests where:
1. predicate is not a function
2. predicate is not provided
3. function returns false

**Review by Planeshifter — APPROVED** [review]
LGTM! Thanks for this contribution!

**Pull request merged**